### PR TITLE
Support noise model in simulator construction

### LINKED ISSUE #320: Consider adding `noise` parameter to QsimSimulator constructor
`cirq.DensityMatrixSimulator` includes a [`noise`](https://quantumai.google/reference/python/cirq/sim/DensityMatrixSimulator) parameter in its constructor. Consider adding the constructor to `QsimSimulator` as well now that quantum trajectories are in place.

On the other hand, the noise parameter is easily processed with one line:
```
noisy_circuit = noise.noisy_moments(circuit, sorted(circuit.all_qubits()))
```

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Fixes #320. Noise models can now be passed to the simulator directly, similar to `cirq.Simulator`; this allows the qsim simulator to be used with Cirq experiments that take a `Sampler` to mimic hardware.

CC @verult

### CODE REVIEW

File: qsimcirq/qsim_simulator.py [hunk 1]
```diff
@@ -160,8 +164,12 @@ def _sample_measure_results(
             ValueError: If there are multiple MeasurementGates with the same key,
                 or if repetitions is negative.
         """
-        if not isinstance(program, qsimc.QSimCircuit):
-            program = qsimc.QSimCircuit(program, device=program.device)
+
+        # Add noise to the circuit if a noise model was provided.
+        program = qsimc.QSimCircuit(
+            self.noise.noisy_moments(program, program.all_qubits()),
```
Comment (verult) [inline qsimcirq/qsim_simulator.py:170]: Why are qubits `sorted()` in the base implementation but not here?

**Comment by verult** [comment]
lgtm, just one minor comment

File: qsimcirq/qsim_simulator.py (hunk 1 above)
Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py]: Good call, thanks!

**Comment by 95-martin-orion** [comment]
pinging @MichaelBroughton for review (I don't think @verult's approval allows me to merge)

**Comment by 95-martin-orion** [comment]
This has been sitting around since I forgot about it, but AFAICT there's no reason we shouldn't be able to merge it.

File: qsimcirq/qsim_simulator.py [hunk 2]
```diff
@@ -303,8 +307,12 @@ def _sample_measure_results(
             ValueError: If there are multiple MeasurementGates with the same key,
                 or if repetitions is negative.
         """
-        if not isinstance(program, qsimc.QSimCircuit):
-            program = qsimc.QSimCircuit(program, device=program.device)
+
+        # Add noise to the circuit if a noise model was provided.
+        program = qsimc.QSimCircuit(
+            self.noise.noisy_moments(program, sorted(program.all_qubits())),
```
Comment (sergeisakov) [inline qsimcirq/qsim_simulator.py:313]: Here and in three places below, `program.all_qubits()` can be called only once. It seems there is also some additional overhead if no noise model is passed to the simulator.

**Review by sergeisakov — CHANGES_REQUESTED** [review]

**Review by sergeisakov — APPROVED** [review]
LGTM

**Pull request merged**